### PR TITLE
BUG: Fix assert_equal when check_exact=True for non-numeric dtypes #3…

### LINKED ISSUE #35446: BUG: pd.testing.assert_frame_equal(..., check_exact=True) raises assertion error if any columns are not numeric
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
import pandas as pd
a = pd.DataFrame({"a": list("bc"), "b": [1, 2]})
pd.testing.assert_frame_equal(a, a.copy(), check_exact=True)
```

```pytb
   ...: a = pd.DataFrame({"a": list("bc"), "b": [1, 2]}) 
   ...: pd.testing.assert_frame_equal(a, a.copy(), check_exact=True)                                                               
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-2-bac78c400991> in <module>
      1 import pandas as pd
      2 a = pd.DataFrame({"a": list("bc"), "b": [1, 2]})
----> 3 pd.testing.assert_frame_equal(a, a.copy(), check_exact=True)

    [... skipping hidden 2 frame]

AssertionError: check_exact may only be used with numeric Series
```

#### Problem description

Previously, the example above would not have thrown an error. It makes it a less useful function if I have to start subdividing dataframes by types just to check that they are equal. Presumably check_exact should either have no effect if called on a non-numeric `Series` or `assert_frame_equal` should not pass `check_exact=True` if the series to be compared isn't numeric.

It looks like this was introduced recently, in https://github.com/pandas-dev/pandas/commit/08c65973643b73e877133d1a7391de7092f4b267

#### Expected Output

I expected it to work like it did previously, and not throw an error. [Here's an example of test failures this is causing](https://travis-ci.org/github/theislab/anndata/jobs/712782637).

#### Output of ``pd.show_versions()``

<details>

```
INSTALLED VERSIONS
------------------
commit           : d9fff2792bf16178d4e450fe7384244e50635733
python           : 3.8.5.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 19.5.0
Version          : Darwin Kernel Version 19.5.0: Tue May 26 20:41:44 PDT 2020; root:xnu-6153.121.2~2/RELEASE_X86_64
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.1.0
numpy            : 1.19.0
pytz             : 2020.1
dateutil         : 2.8.1
pip              : 20.1.1
setuptools       : 49.2.0
Cython           : None
pytest           : 5.4.3
hypothesis       : None
sphinx           : 3.1.2
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.11.2
IPython          : 7.16.1
pandas_datareader: None
bs4              : None
bottleneck       : None
fsspec           : 0.7.4
fastparquet      : None
gcsfs            : None
matplotlib       : 3.3.0
numexpr          : 2.7.1
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : 0.17.1
pytables         : None
pyxlsb           : None
s3fs             : None
scipy            : 1.5.1
sqlalchemy       : 1.3.18
tables           : 3.6.1
tabulate         : None
xarray           : 0.16.0
xlrd             : 1.2.0
xlwt             : None
numba            : 0.50.1
```

</details>

### PR DESCRIPTION
**Pull request by ivirshup**
`assert_series_equal(..., check_exact=True)` no longer raises an error when series has a non-numeric dtype. This fixes a bug/ regression introduced in `1.1.0`.

- [x] closes #35446
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

I've added a very simple test, but could add more. For instance, there could be a check on `assert_frame_equal`, but I'm not sure what the desired strategy is for that.

I've left off the `whatsnew` entry for now since I figure that's the most likely to generate conflicts while this gets reviewed.

### CODE REVIEW

**Review by jreback — CHANGES_REQUESTED** [review]
looks good @ivirshup; can you add a note in the 1.1.1 docs (regression section). ping on green.

**Comment by ivirshup** [comment]
@jreback, added the release note, and it's kinda green. It looks like one of the jobs timed out though.

**Comment by simonjayhawkins** [comment]
> It looks like one of the jobs timed out though.

restarted

File: pandas/tests/util/test_assert_series_equal.py [hunk 1]
```diff
@@ -281,3 +281,13 @@ class MySeries(Series):
...
         tm.assert_series_equal(s3, s1, check_series_type=True)
+
+
+def test_series_equal_exact_for_nonnumeric():
+    # https://github.com/pandas-dev/pandas/issues/35446
+    s1 = Series(["a", "b"])
+    s2 = Series(["a", "b"])
+    s3 = Series(["b", "a"])
+
+    _assert_series_equal_both(s1, s2, check_exact=True)
```
Comment (jreback) [inline pandas/tests/util/test_assert_series_equal.py:292]: can you just use tm.assert_series_equal rather than the internal function

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
also pls merge master

File: pandas/tests/util/test_assert_series_equal.py (hunk 1 above)
Comment (ivirshup) [reply pandas/tests/util/test_assert_series_equal.py]: 👍

**Comment by ivirshup** [comment]
Made those changes. How is it now?

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks!

**Comment by jreback** [comment]
@meeseeksdev backport 1.1.x

**Pull request merged**